### PR TITLE
[#66] Figure out authentication for GitHub App and use proper token for submitting comments

### LINKED ISSUE #66: Figure out authentication for GitHub App and use proper token for submitting comments

### PR DESCRIPTION
**Pull request by chshersh**
Resolves #66

This huge PR implements authentification with GitHub Applications. It works, I was able to submit a review with `hint-man`:

* https://github.com/kowainik/hintman-target/pull/2#pullrequestreview-276246163

The implementation is dirty, and it's not complete... But further cleanup and wiring up can be done later.

The idea is the following:

1. For each `InstallationId` generate an access token that can be used for requests.
2. When APP is installed, remember generated access `GitHubToken` for all `Owner/Repo` pairs for which the app installed. 

Now, in the PR events, one can get `GitHubToken` from `TokenCache` and use it to submit reviews.

Please, feel free to share any feedback regarding the implementation!

### CODE REVIEW

File: src/Hintman/HLint.hs [hunk 1]
```diff
@@ -56,6 +56,7 @@ getFileHLintSuggestions (fileName, content) = case content of
 -- | In-memory file download from GitHub PR sources by given URL.
 downloadFile :: MonadIO m => Text -> m (Maybe ByteString)
 downloadFile url = do
+    -- TODO: put in the context to not create each time
```
Comment (vrom911) [inline src/Hintman/HLint.hs]: :+1:

File: src/Hintman/Effect/TokenStorage.hs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    lookupToken = lookupTokenImpl
+
+insertTokenImpl
+    :: (MonadReader env m, Has TokenCache env, MonadIO m)
+    => Owner
+    -> Repo
+    -> GitHubToken
+    -> m ()
+insertTokenImpl owner repo key = do
+    putTextLn $ "Inserting token for: " <> show owner <> "/" <> show repo
```
Comment (vrom911) [inline src/Hintman/Effect/TokenStorage.hs]: We need to add logging...

File: src/Hintman/Installation.hs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+import Data.Time.Clock.POSIX (utcTimeToPOSIXSeconds)
+import Network.HTTP.Client (Request (..), Response (..), httpLbs, parseUrlThrow)
+import Network.HTTP.Client.TLS (newTlsManager)
+import Network.HTTP.Types (Status (..))
+import Web.JWT (JWTClaimsSet (..), Signer (RSAPrivateKey), encodeSigned, numericDate, stringOrURI)
+
+import Hintman.Core.Token (AppInfo (..), InstallationAccessToken, InstallationId (..))
+
+
+newtype JwtToken = JwtToken Text
```
Comment (vrom911) [inline src/Hintman/Installation.hs]: Shouldn't we move this to `Hintman.Core.Jwt`?

File: src/Hintman/Webhook.hs [hunk 4]
```diff
@@ -5,37 +5,110 @@ module Hintman.Webhook
...
 
 hintmanServer :: HintmanSite (AsServerT App)
 hintmanServer = HintmanSite
     { hintmanListenRoute = issueCommentHook
+    , hintmanInstallAppRoute = appInstalledHook
+    , hintmanInstallRepoRoute = repoInstalledHook
     }
 
 issueCommentHook :: MonadIO m => RepoWebhookEvent -> ((), PullRequestEvent) -> m ()
 issueCommentHook _ ev = print $ evPullReqPayload $ snd ev
```
Comment (vrom911) [inline src/Hintman/Webhook.hs]: Forgot to rename that... Sorry!

File: src/Hintman/Webhook.hs [hunk 5]
```diff
@@ -5,37 +5,110 @@ module Hintman.Webhook
...
-newtype HintmanSite route = HintmanSite
-    { hintmanListenRoute :: route
+data HintmanSite route = HintmanSite
+    { -- | Main endpoint that listens PR events.
+      hintmanListenRoute :: route
         :- GitHubEvent '[ 'WebhookPullRequestEvent ]
         :> GitHubSignedReqBody '[JSON] PullRequestEvent
         :> Post '[JSON] ()
+
+      -- | Triggered when a GitHub App has been installed or uninstalled.
```
Comment (vrom911) [inline src/Hintman/Webhook.hs]: What GitHub App is meant here? Is it about Hintman itself only?

File: src/Hintman/Effect/TokenStorage.hs (hunk 2 above)
Comment (chshersh) [reply src/Hintman/Effect/TokenStorage.hs]: Yes, we really need #39...

File: src/Hintman/Installation.hs (hunk 3 above)
Comment (chshersh) [reply src/Hintman/Installation.hs]: This data type is internal and exists only to improve readability. The problem is that `encodeSigned` function from `jwt` library returns `Text`:

* http://hackage.haskell.org/package/jwt-0.10.0/docs/Web-JWT.html#v:encodeSigned

I feel uncomfortable using plain `Text` for `JwtToken`. This data type is just an internal implementation detail and it's used by the exposed interface, so I didn't bother to put it in a separate module...

File: src/Hintman/Webhook.hs (hunk 4 above)
Comment (chshersh) [reply src/Hintman/Webhook.hs]: I didn't notice :smile:

File: src/Hintman/Webhook.hs (hunk 5 above)
Comment (chshersh) [reply src/Hintman/Webhook.hs]: Yes, it's about Hintman App. This event happens when you are trying to install Hintman via GitHub web interface. I can rename `GitHub App` to `Hintman GitHub App` or `Hintman` for readability

**Pull request merged**